ipaclient/plugins/baseuser.py

```diff
@@ -0,0 +1,84 @@
+#
+# Copyright (C) 2022  FreeIPA Contributors see COPYING for license
+#
+
+import locale
+import logging
+import subprocess
+from ipaclient.frontend import MethodOverride
+from ipalib import errors
+from ipalib import Bool, Flag, StrEnum
+from ipalib.text import _
+
+logger = logging.getLogger(__name__)
+
+
+class baseuser_add_passkey(MethodOverride):
+    takes_options = (
+        Flag(
+            'register',
+            cli_name='register',
+            doc=_('Register the passkey'),
+        ),
+        Bool(
+            'require_user_verification?',
+            cli_name='require_user_verification',
+            doc=_('Require user verification during authentication with '
+                  'the passkey')
+        ),
+        StrEnum(
+            'cosetype?',
+            cli_name='cose_type',
+            doc=_('COSE type to use for registration'),
+            values=('es256', 'rs256', 'eddsa'),
+        ),
+    )
+
+    def get_args(self):
+        # ipapasskey is not mandatory as it can be built
+        # from the registration step
+        for arg in super(baseuser_add_passkey, self).get_args():
+            if arg.name == 'ipapasskey':
+                yield arg.clone(required=False, alwaysask=False)
+            else:
+                yield arg.clone()
+
+    def forward(self, *args, **options):
+        if self.api.env.context == 'cli':
+            # 2 formats are possible for ipa user-add-passkey:
+            # --register [--require-user-verification] [--cose-type ...]
+            # or
+            # passkey:<key id>,<pub key>
+            for option in super(baseuser_add_passkey, self).get_options():
+                if args and option in options:
+                    raise errors.MutuallyExclusiveError(
+                        reason=_("cannot specify both %s and "
+                                 "passkey mapping").format(option))
+            # if the first format is used, need to register the key first
+            # and obtained the data
+            if 'register' in options:
+                options.pop('register')
+                cosetype = options.pop('cosetype', None)
+                require_verif = options.pop('require_user_verification', None)
+                cmd = ["/usr/libexec/sssd/passkey_child", "--register",
```

Comment:
It would be good to add a bit better detection logic that helps with multiple distributions and total absence of the utility:
- move path to passkey_child to ipaplatform paths code
- add detector that probes whether the passkey_child binary exists and executable
- in case of a failure to probe, display warning and only process a passkey string, not a registration.

Reply:
Good point, I added a check